### PR TITLE
TST, MAINT: Make sure exceptions of `inner` and `dot` match for different cases

### PR DESCRIPTION
**Pull request by jakirkham**
Related: https://github.com/numpy/numpy/pull/6988

Fixes an issue with `dot` where it was  `ValueError` or a `TypeError` depending on the order of its arguments. This differs from `inner`, which always raises a `TypeError`. Below is an example of the current behavior, which seems odd. This PR makes all cases raise a `TypeError`.

```
>>> import numpy
>>> import numpy as np
>>> np.inner(1., np.array((1,1), dtype='i,i'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: invalid type promotion
>>> np.inner(np.array((1,1), dtype='i,i'), 1.)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Cannot cast array data from dtype([('f0', '<i4'), ('f1', '<i4')]) to dtype('float64') according to the rule 'safe'
>>> np.dot(1., np.array((1,1), dtype='i,i'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Cannot find a common data type.
>>> np.dot(np.array((1,1), dtype='i,i'), 1.)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Cannot cast array data from dtype([('f0', '<i4'), ('f1', '<i4')]) to dtype('float64') according to the rule 'safe'
```

### CODE REVIEW

File: numpy/core/tests/test_multiarray.py [hunk 1]
```diff
@@ -2041,6 +2041,24 @@ def __numpy_ufunc__(self, ufunc, method, pos, inputs, **kwargs):
         assert_raises(TypeError, np.dot, b, c)
         assert_raises(TypeError, c.dot, b)
 
+    def test_dot_type_mismatch(self):
+        c = 1.
+        A = np.array((1,1), dtype='i,i')
+
+        try:
```
Comment (charris) [inline numpy/core/tests/test_multiarray.py]: Any reason not to use `assert_raises` here?

Comment (jakirkham) [inline numpy/core/tests/test_multiarray.py]: None, I wasn't sure the right way to do this in `numpy`'s test suite and took `test_mul` as an example.

**Comment by charris** [comment]
IIRC, there was a suggestion that `TypeError` would be correct in both cases. Not sure what our policy on changing error types is, we probably shouldn't, but we do on occasion...

**Comment by jakirkham** [comment]
I feel like one of these is a bug, but I am just not sure which one. I agree that `TypeError` sounds reasonable. The only benefit to keeping one as `ValueError` would be to handle the too cases separately, which I am not sure is so valuable.

**Comment by jakirkham** [comment]
FWIW, if we decided changing the exception is just not ok, it would be trivial to preserver the behavior in PR ( https://github.com/numpy/numpy/pull/6968 ), but I just wanted to raise this oddity separately so it could be discussed more directly.

File: numpy/core/tests/test_multiarray.py (hunk 1 above)
Comment (jakirkham) [inline numpy/core/tests/test_multiarray.py]: Would you prefer it changed?

**Comment by jakirkham** [comment]
Another nice consequence of changing everything to `TypeError` is if someone was already using `dot` like so...

    try:
        np.dot(a, b)
    except TypeError, ValueError:
        pass

It will still work.

**Comment by jakirkham** [comment]
Interesting, seeing a failure on debug build. Not sure if this is related ( http://bugs.python.org/issue20500 ).

File: numpy/core/tests/test_multiarray.py (hunk 1 above)
Comment (jakirkham) [inline numpy/core/tests/test_multiarray.py]: It was changed to `assert_raises`.

**Comment by charris** [comment]
I can reproduce this here with 3.4.4. A bit more info is in the comment in the cpython code.
```
#ifdef Py_DEBUG
    /* PyObject_Repr() must not be called with an exception set,
       because it may clear it (directly or indirectly) and so the
       caller loses its exception */
    assert(!PyErr_Occurred());
#endif
```

**Comment by jakirkham** [comment]
After looking a bit closer and seeing what you said here @charris, I think this is the issue ( https://github.com/numpy/numpy/issues/6990 ).

**Comment by jakirkham** [comment]
Failures on AppVeyor ( https://github.com/numpy/numpy/issues/6991 ) had not previously occurred for this PR with exception of the segmentation fault issue (also happened on Travis), which has since been resolved. I believe these to be unrelated to the content of this PR.

**Comment by jakirkham** [comment]
As AppVeyor is merging with master, which is broken, it is currently failing the tests there. So, I ran my own AppVeyor build without this merge (this is rebased on a commit on `master` before the bad commit) and it looks like it checks out ( https://ci.appveyor.com/project/jakirkham/numpy/build/1.0.5 ).

**Comment by charris** [comment]
@jakirkham Needs a rebase.

**Comment by jakirkham** [comment]
Sure, @charris, can do. Are we actually happy with this change though or should we switch everything to TypeError? Should we get some more people to weigh in?

**Comment by charris** [comment]
I certainly wouldn't mind more feedback. @njsmith @seberg Thoughts?

**Comment by jakirkham** [comment]
I rebased it so hopefully the proposed change is clearer to see.

**Comment by jakirkham** [comment]
Honestly, I kind of think they all of them should be `TypeError`s.

**Comment by charris** [comment]
Yeah, lets go with `TypeError`. That should be mentioned in the 1.11.0 release notes under compatibility.

**Comment by jakirkham** [comment]
Sure, I'll updated. There are a couple other cases in the same file where it raises a `ValueError` when it probably should be a `TypeError`, but we can look at those later.

File: doc/release/1.11.0-notes.rst [hunk 2]
```diff
@@ -174,6 +174,12 @@ This behaviour mimics that of other functions such as ``np.diagonal`` and
 ensures, e.g., that for masked arrays ``np.trace(ma)`` and ``ma.trace()`` give
 the same result.
 
+*np.dot* raises ``TypeError`` instead of ``ValueError``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This behaviour mimics that of other functions such as ``np.inner``. If the two
+arguments cannot be cast to a common type, ``np.dot`` will now always raise a
+``TypeError``. Before it had raise ``ValueError`` in some cases.
```
Comment (njsmith) [inline doc/release/1.11.0-notes.rst]: had -> could

**Comment by njsmith** [comment]
> I certainly wouldn't mind more feedback. @njsmith @seberg Thoughts?

Sorry, missed this. I don't have a strong opinion either way :-).

**Comment by jakirkham** [comment]
No worries. It was a pretty minor one. Thanks for the feedback.

File: doc/release/1.11.0-notes.rst (hunk 2 above)
Comment (jakirkham) [inline doc/release/1.11.0-notes.rst]: I realized after rereading this there were many issues and so rewrote it. Also, is this in the right place? I wasn't sure where this sort of change went so just stuck it here.

**Comment by seberg** [comment]
Well, I would not worry about changing this at all. Changing it the other way around (dot) maybe a little, but also not too much, if you feel strongly that the current error type is bad, I think I won't stop you.

**Comment by charris** [comment]
I'd put iot under compatibility notes rather than changes.

**Comment by charris** [comment]
Or maybe both ;)

**Comment by jakirkham** [comment]
@charris, it is now in both places. :)

@seberg, yeah, I think having two different types of exceptions for the same thing is weird. Then again, maybe I'm just picky.

**Comment by charris** [comment]
Thanks @jakirkham .

**Comment by jakirkham** [comment]
Sure. Thanks everyone.

**Pull request merged**